### PR TITLE
PREEMPT_RT support for EVE

### PR DESCRIPTION
**Pull request by rouming**
This PR introduces the PREEMPT_RT support for the main EVE kernel (pkg/kernel) for the x86_64 architecture by

0. Bumping up the minor version of the main kernel to the 5.1.131 to be aligned with PREEMPT_RT patches.
    
1. Adding the official `patch-5.10.131-rt72.patch.gz` set of patches.
    
2. Adding the patch which does not export the migrate_enable() and migrate_disable() kernel calls as GPL. OpenZFS is not GPL but uses those calls. Before the PREEMPT_RT patches those two calls were declared as "static inline" in the header, thus no export was needed. 
    
3. Once the 'make pkg/kernel PREEMPT_RT=1' is specified the PREEMPT_RT kernel flavour is built in the 'eve-rt-kernel' docker image.
    
4. To build a live image with the PREEMPT_RT kernel call as the following: 'make live PREEMPT_RT=1'. The same applies for the installer images, e.g.: 'make installer-iso PREEMPT_RT=1'.

The PR is quite conservative and introduces a new kernel flavor which can be enabled on demand, so that by default other functionality is not touched.

This PR targets only amd64, since PREEMPT_RT scheduler can't be enabled on arm64 along with the KVM support (what a bummer).

There are a few changes in Dockerfile (recent security improvements) and kernel-build-yml.sh script, so @shjala please take a look.

### CODE REVIEW

File: pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch
Comment (rene) [inline pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch:53]: I didn't notice we were using SLAB before.... but I guess you are changing to SLUB only for improvements, right?

File: pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch
Comment (rene) [inline pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch:168]: Enable DEBUG here will increase overhead and affect the performance, I don't think we should enabled by default...

File: pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch
Comment (rouming) [reply pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch:53]: This is the default, comes with the preemt_rt. I left untouched.

File: pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch
Comment (rouming) [reply pkg/kernel/kernel-config/kernel_config-5.10.x-preempt-rt-x86_64.patch:168]: Left as is intentionally. Want to catch any complains from kernel if preempted in unsafe way. Want to see the whole virtualization stack in action. Will disable later.

File: pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch
Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:22]: Before those methods were declared as static inline in the header, so no export beeded

Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:22]: *needed

File: pkg/kernel/Dockerfile
Comment (eriknordmark) [inline pkg/kernel/Dockerfile:61]: Can you do this version bump as a separate commit to make it easier to see in git log?

File: pkg/kernel/Dockerfile
Comment (eriknordmark) [inline pkg/kernel/Dockerfile:13]: This is a string, right? So shouldn't the default be an empty string?

File: pkg/kernel/Dockerfile
Comment (rouming) [reply pkg/kernel/Dockerfile:61]: Already in a separate commit: pkg/kernel: bump up minor version of the current stable kernel to 5.1… ￼…

File: pkg/kernel/Dockerfile
Comment (rouming) [reply pkg/kernel/Dockerfile:13]: We need the default value to satisfy sbom, which fails if nothing is defined, because there is a FROM line:
```
FROM kernel-build-preempt-rt-${PREEMPT_RT_SUPPORT} AS kernel-build
```

**Comment by rouming** [comment]
> 

Frankly I don't want to bother with a new kernel now, because it is used for risc only and the future of the whole new-kernel is not clear: do we drop it and switch the main kernel to a 5.15?

File: Makefile
Comment (eriknordmark) [reply Makefile:32]: Which values are allowed for this?
It seems like it can be one of three things: true, false, or "--preempt-rt-support"??
But the Dockerfile expects true or false?
Something is confusing to me.

Comment (eriknordmark) [reply Makefile:32]: Which values are allowed for this?
It seems like it can be one of three things: true, false, or "--preempt-rt-support"??
But the Dockerfile expects true or false?
Something is confusing to me.

**Comment by eriknordmark** [comment]
> > Should we do a separate PR which applies this to new-kernel as well?
> 
> @eriknordmark Frankly I don't want to bother with a new kernel now, because it is used for risc only and the future of the whole new-kernel is not clear: do we drop it and switch the main kernel to a 5.15?

Over the next few months we should update the kernel. But if we keep them in synch that can be done by "rm -rf kernel; mv new-kernel kernel; mkdir new-kernel; // fill new-kernel with next kernel LTS"

Don't know if it is worth the effort to keep them in synch for that reason.
@rene  any thoughts on this?

> 
> Update: or better to invest time and unify both Dockerfile and do the same trick with building a 5.15 on demand by 'KERNEL=5.15' make argument?

Probably not worth while if we are going to externalize the kernel build from the repo.

File: pkg/kernel/Dockerfile
Comment (eriknordmark) [reply pkg/kernel/Dockerfile:13]: OK

File: Makefile
Comment (rouming) [reply Makefile:32]: Seems the github is indeed broken today, I experience the same two-times-comments.

Anyway, this "--preempt-rt-support" is an argument for the build-kernel-yaml.sh script, which sets the build dockerfile argument, the PREEMPT_RT_SUPPORT=true.

Comment (rouming) [reply Makefile:32]: The chain is the following
Makefile -> kernel-build-yml.sh -> build.yml -> Dockerfile

**Comment by rouming** [comment]
> Probably not worth while if we are going to externalize the kernel build from the repo.

This does not eliminate the dockerfile and all the efforts to keep both in sync, we just skip the step of downloading the kernel sources and applying patches. Other than that stays the same.

File: Makefile
Comment (eriknordmark) [reply Makefile:32]: It sure would be less confusing if in the Makefile this was called KERNEL_BUILD_YML_ARG or something that that.

File: tools/kernel-build-yml.sh
Comment (eriknordmark) [inline tools/kernel-build-yml.sh:22]: Does this mean you'll push an image of that name to docker hub? (If so we need to ask IT to create that under lfedge.)

Comment (rouming) [reply tools/kernel-build-yml.sh:22]: I build the image with this name, true. But when the image is pushed - is not clear to me. Do i need to do that explicitly? Will shoot a ticket on lf-edge Jira to create a corresponding docker package.

File: Makefile
Comment (rouming) [reply Makefile:32]: This is very subjective :) for example i clearly memorize variables much better if they consist of max 3 nouns, more than that for me is usually garbage which is difficult to remember, like "kernel build yml arg" or "build kernel yml arg", or other combinations. 

But on this particular matter i don't have any preferences, since this is very minor,  I will update the PR, no problem.

File: Makefile
Comment (uncleDecart) [inline Makefile:385]: Do I get it right, that now kernel-yml will always be quiet?

**Comment by uncleDecart** [comment]
@rouming you say this PR targets only amd64, do we need to include some checks that PREEMT_RT will not build for arm or did I miss them?

File: tools/kernel-build-yml.sh
Comment (rouming) [reply tools/kernel-build-yml.sh:22]: Left the ticket on LFEDGE to create the 'eve-rt-kernel' repo.

File: Makefile
Comment (rouming) [reply Makefile:385]: It seems nothing has been changed, QUIET was always there.

**Comment by rouming** [comment]
> @rouming you say this PR targets only amd64, do we need to include some checks that PREEMT_RT will not build for arm or did I miss them?

The config for other arches is missing, so no worries, you get an error :)

File: Makefile
Comment (uncleDecart) [reply Makefile:385]: I guess I like python more than bash :D But nevertheless, you are removing check if `pkg/kernel/build.yml` and `pkg/new-kernel/build.yml` files exist and instead we always put `tools/kernel-build-yml.sh` with PREEMT_RT variable. Why do we skip it now?

**Comment by rouming** [comment]
Difference to the previous version:
* s/PREEMPT_RT_SUPPORT/PREEMPT_RT_SUPPORT_ARG/ in the Makefile

File: Makefile
Comment (rouming) [reply Makefile:385]: The commit message should tell the whole story: https://github.com/lf-edge/eve/pull/3254/commits/e66a85ad6feeb2086aa350435d1e7791ecc0fcd5
What's probably is missing in the description is the following line: "We need to generate the build.yml each time to reflect the changes in the PREEMPT_RT_SUPPORT build argument"

**Comment by rouming** [comment]
Difference to the previous version:
* Just commit message tweaks

**Comment by rouming** [comment]
I pay 5 euro for approval. Offer is valid today only.

**Comment by rene** [comment]
> > > Should we do a separate PR which applies this to new-kernel as well?
> > 
> > 
> > @eriknordmark Frankly I don't want to bother with a new kernel now, because it is used for risc only and the future of the whole new-kernel is not clear: do we drop it and switch the main kernel to a 5.15?
> 
> Over the next few months we should update the kernel. But if we keep them in synch that can be done by "rm -rf kernel; mv new-kernel kernel; mkdir new-kernel; // fill new-kernel with next kernel LTS"
> 
> Don't know if it is worth the effort to keep them in synch for that reason. @rene any thoughts on this?
> 
> > Update: or better to invest time and unify both Dockerfile and do the same trick with building a 5.15 on demand by 'KERNEL=5.15' make argument?
> 
> Probably not worth while if we are going to externalize the kernel build from the repo.

They are not in that sync anyways, I don't think is worth to introduce all these change in new-kernel right now....

File: pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch
Comment (christoph-zededa) [inline pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: Are we sure that from a law/license point of view this is allowed?

Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: Until we don't ship the product - why not?

Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: Or even we ship the product

**Review by mikem-zed — APPROVED** [review]
LGTM after the issue with docker tag is resolved

**Comment by rouming** [comment]
Difference to the previous version:
* Docs tweaks

File: pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch
Comment (christoph-zededa) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: Might be worth to ask a lawyer ...

Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: If this raises questions I can drop the compilation of the openzfs for preempt_rt.

Comment (rouming) [reply pkg/kernel/patches-preempt-rt-5.10.x/0001-sched-core-make-migrate_disable-and-migrate_enable-a.patch:23]: @eriknordmark interesting question from Christoph

**Comment by shjala** [comment]
LGTM

File: tools/kernel-build-yml.sh
Comment (eriknordmark) [reply tools/kernel-build-yml.sh:22]: FWIW the publish workflow is what pushes to docker hub. I haven't checked it only pushes very specific things like lfedge/eve:<tag>, or other containers.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Pull request merged**
